### PR TITLE
Use base timers in native ledger

### PR DESCRIPTION
**Pull request by zenparsing**
Resolves https://github.com/brave/brave-browser/issues/8676

## Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: vendor/bat-native-ledger/src/bat/ledger/internal/common/time_util.cc [hunk 1]
```diff
@@ -33,4 +35,23 @@ uint64_t GetCurrentTimeStamp() {
...
+  uint64_t seconds = brave_base::random::Geometric(delay.InSecondsF());
+  return base::TimeDelta::FromSeconds(static_cast<int64_t>(seconds));
+}
+
+base::TimeDelta GetRandomizedDelayWithBackoff(
+    base::TimeDelta delay,
+    base::TimeDelta max_delay,
+    int backoff_count) {
+  for (int i = 0; i < backoff_count; ++i) {
+    delay *= 2;
```
Comment (tmancey) [inline vendor/bat-native-ledger/src/bat/ledger/internal/common/time_util.cc:48]: nit: use shift operator `delay << backoff_count`

**Comment by tmancey** [comment]
@zenparsing you may want to check https://github.com/brave/brave-core/blob/master/vendor/bat-native-ads/src/bat/ads/internal/retry_timer.cc, thanks

File: vendor/bat-native-ledger/src/bat/ledger/internal/common/time_util.cc (hunk 1 above)
Comment (zenparsing) [reply vendor/bat-native-ledger/src/bat/ledger/internal/common/time_util.cc:48]: `delay` here is a `base::TimeDelta`, which doesn't support overloading for shift operators (other than `ostream` of course).

Also, with shifting we'd need to be careful to not shift too far (we'd have to range check `backoff_count`).

Comment (tmancey) [reply vendor/bat-native-ledger/src/bat/ledger/internal/common/time_util.cc:48]: but we could multiple delay by `1 << backoff_count` to remove the loop. and cap backoff_count to a maximum delay

Comment (zenparsing) [reply vendor/bat-native-ledger/src/bat/ledger/internal/common/time_util.cc:48]: nice @tmancey - will do.

File: vendor/bat-native-ledger/src/bat/ledger/internal/api/api_parameters.cc [hunk 2]
```diff
@@ -97,23 +92,25 @@ void APIParameters::RunCallbacks() {
...
-    start_in = base_delay + random_delay;
-  }
+  base::TimeDelta start_in =
+      base_delay + braveledger_time_util::GetRandomizedDelay(delay);
+
+  BLOG(1, "Params timer set for " << start_in);
 
-  BLOG(1, "Params timer start in " << start_in);
+  refresh_timer_.Start(FROM_HERE, start_in,
+      base::BindOnce(&APIParameters::OnTimerElapsed, base::Unretained(this)));
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/internal/api/api_parameters.cc:109]: can we just call `Fetch` directly here so that we don't need another function `OnTimerElapsed`

**Comment by zenparsing** [comment]
CI failed on unrelated `windows:GreaselionServiceTest.ScriptInjectionWithPrecondition`.

File: vendor/bat-native-ledger/src/bat/ledger/internal/contribution/contribution.cc [hunk 3]
```diff
@@ -655,18 +624,15 @@ void Contribution::OnResult(
   }
 
   if (result == ledger::Result::RETRY_LONG) {
-    if (ledger::short_retries) {
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/internal/contribution/contribution.cc:658]: why was this removed?

File: vendor/bat-native-ledger/src/bat/ledger/internal/api/api_parameters.cc [hunk 4]
```diff
@@ -67,21 +57,22 @@ void APIParameters::OnFetch(const ledger::UrlResponse& response) {
       braveledger_response_util::ParseParameters(response, &parameters);
   if (result == ledger::Result::RETRY_SHORT) {
     RunCallbacks();
-    SetRefreshTimer(brave_base::random::Geometric(90));
+    SetRefreshTimer(base::TimeDelta::FromSeconds(90));
```
Comment (emerick) [inline vendor/bat-native-ledger/src/bat/ledger/internal/api/api_parameters.cc:60]: Is there a reason to switch away from the geometric time? Was it just never necessary for these timers?

Comment (zenparsing) [reply vendor/bat-native-ledger/src/bat/ledger/internal/api/api_parameters.cc:60]: Thanks for the double-check on this! The randomization is applied to these values in `SetRefreshTimer`.

Comment (emerick) [reply vendor/bat-native-ledger/src/bat/ledger/internal/api/api_parameters.cc:60]: Got it!
